inbm/dispatcher-agent/dispatcher/sota/oras_util.py

```diff
@@ -0,0 +1,150 @@
+        raise DispatcherException(err)
+
+    logger.debug(f"source: {source}, "
+                 f"registry_server: {registry_server}, "
+                 f"image: {image}, "
+                 f"image_tag: {image_tag}, "
+                 f"repository_name: {repository_name}, "
+                 f"registry_manifest: {registry_manifest}")
+
+    verify_source(source=source, dispatcher_broker=dispatcher_broker)
+    dispatcher_broker.telemetry('Source Verification check passed')
+
+    try:
+        enough_space = is_enough_space_to_download(
+            registry_manifest, repo, password)
+    except DispatcherException as e:
+        raise DispatcherException(e)
+
+    if not enough_space:
+        err_msg = " Insufficient free space available on " + repo.get_repo_path() + \
+                  " for " + str(uri.value)
+        raise DispatcherException(err_msg)
+
+    if password:
+        logger.debug("JWT Token provided.")
+    else:
+        err_msg = " No JWT token. Abort the update. "
+        raise DispatcherException(err_msg)
+
+    msg = f'Fetching software package from {image_full_path}'
+    dispatcher_broker.telemetry(msg)
+
+    # Call oras to pull the image
+    # The password is the JWT token
+    (out, err_run, code) = PseudoShellRunner().run(f"oras pull {image_full_path} -o {repo.get_repo_path()} --password {password}")
+    if code != 0:
+        if err_run:
+            raise DispatcherException("Error to download OTA files with ORAS: " + err_run + ". Code: " + str(code))
+        else:
+            raise DispatcherException("Error to download OTA files with ORAS. Code: " + str(code))
+    else:
+        dispatcher_broker.telemetry('OTA Download Successful.')
+
+
+def is_enough_space_to_download(manifest_uri: str,
+                                destination_repo: IRepo,
+                                jwt_token: str) -> bool:
+    """Checks if enough free space exists on platform to hold download.
+
+    Calculates the file size from the OCI server and checks if required free space is available on
+    the platform.
+    @param manifest_uri: registry manifest uri
+    @param destination_repo:  desired download destination
+    @param jwt_token: jwt_token provided for access the release server
+    """
+    try:
+        logger.debug(f"Checking OCI artifact size with manifest uri: {manifest_uri}")
+        headers = {
+            "Authorization": f"Bearer {jwt_token}",
+            "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json"
+        }
+        response = requests.get(manifest_uri, headers=headers)
+        data = json.loads(response.text)
+        logger.debug(f"resp={data}")
+        # Calculate the total size
+        file_size = 0
+        for layer in data['layers']:
```

Comment:
accessing keys in `data` could raise exceptions if they don't exist--check key first, or handle exceptions

Reply:
KeyError is added in try and exception statement for error handling.